lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -1101,3 +1117,39 @@ Value Context::materializeConversion(Type type, Value value, bool isSigned,
     value = builder.create<moore::ConversionOp>(loc, type, value);
   return value;
 }
+
+void Context::collectConditionsForUnpackedArray(
+    const slang::ast::FixedSizeUnpackedArrayType &slangType,
+    Value upackedArrayValue, SmallVector<Value> &conditions, Value lhs,
+    Location loc) {
+  Value cond;
+  auto type = convertType(slangType);
+  if (!type) {
+    mlir::emitError(loc, "can't convert slang::ast::FixedSizeUnpackedArrayType "
+                         "to moore::UnpackedArrayType");
+  }
+  auto mooreType = dyn_cast<moore::UnpackedArrayType>(type);
+  const auto &elementType = slangType.elementType;
+  for (slang::int32_t i = slangType.getFixedRange().lower();
+       i <= slangType.getFixedRange().upper(); i++) {
+    auto elemValue = builder.create<moore::ExtractOp>(
+        loc, mooreType.getElementType(), upackedArrayValue, i);
+    if (elementType.isUnpackedArray()) {
+      collectConditionsForUnpackedArray(
+          elementType.as<slang::ast::FixedSizeUnpackedArrayType>(), elemValue,
+          conditions, lhs, loc);
+    } else if (elementType.isSingular()) {
+      if (elementType.isIntegral()) {
+        cond = builder.create<moore::WildcardEqOp>(loc, lhs, elemValue);
+      } else {
+        cond = builder.create<moore::EqOp>(loc, lhs, elemValue);
+      }
+      conditions.push_back(cond);
```

Comment:
From the comment above, the cond can sink towards here.  Or be eliminated by switching to a trinary instead of a if.  Or elminated by duplicating the push_back.